clash-lib/src/Clash/Normalize/Transformations/Specialize.hs

```diff
@@ -234,11 +234,11 @@ constantSpec ctx@(TransformContext is0 tfCtx) e@(App e1 e2)
          else do
            -- Parts of e2 are constant
            let is1 = extendInScopeSetList is0 (fst <$> csrNewBindings specInfo)
-           Letrec newBindings
-            <$> specializeNorm
-                  (TransformContext is1 tfCtx)
-                  (App e1 (csrNewTerm specInfo))
+           body <- specializeNorm
+             (TransformContext is1 tfCtx)
+             (App e1 (csrNewTerm specInfo))
 
+           changed (Letrec newBindings body)
```

Comment:
This is surprising, `specializeNorm` already does a `setChanged` in all cases except when it doesn't find the binding for the global variable it wants to specialize on.

I would say that when `specializeNorm` doesn't do a `setChanged`, then we don't want do `constantSpec` at all. i.e. this code should probably do a `Writer.listen` and use that as a predicate on whether it does a `changed (Letrec newBindings body)` or a `return e`.

Reply:
I think `constantSpecInfo` says that there are bindings it can make, but if the body of that resulting `letrec` is an application on a top entity it won't be specialised by `specialise` so `changed` is never set:

https://github.com/clash-lang/clash-compiler/blob/master/clash-lib/src/Clash/Rewrite/Util.hs#L756

I was still seeing the failure in some tests without adding the `changed` here because

```
Not specializing TopEntity: topEntity

                <no location info>: error:
                    Clash error call:
                    Clash.Rewrite.Util(279): Expression changed without notice(constantSpec): beforetopEntity[GlobalId]
                      (C:(%,,%) testBench_clk[LocalId]
                         (resetGenN removedArg removedArg (SNat 1))
                         (Enable True))
                    after:
                    letrec
                      testBench_done_bindCsr
                      = testBench_clk[LocalId]
                      testBench_done_bindCsr
                      = resetGenN removedArg removedArg (SNat 1)
                    in topEntity[GlobalId]
                         (C:(%,,%) testBench_done_bindCsr[LocalId]
                            testBench_done_bindCsr[LocalId]
                            (Enable True))
                    CallStack (from HasCallStack):
                      error, called at src/Clash/Rewrite/Util.hs:279:5 in clash-lib-1.5.0-inplace:Clash.Rewrite.Util
```

Reply:
Reading up on why we created the let-bindngs: it was to allow constant-specialisation on expression that are in WHNF but not NF due to some FVs. But if we're not specializing to begin with, we don't need to create those let-bindings either. So I still think we should do a `Writer.listen` on `specializeNorm` and only create the let-bindings when we actually specialize.